File: Disciplinas/AED/AED2/slides/u03-Ordenacao_Interna/Selection/selectionAnalysis.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Struct Array
typedef struct Array {
int array[10];
int n;
} Array;

void readArray(Array* vet) {
        printf("Insira %d os elementos: ", vet->n);
        for(int i = 0; i < vet->n; i++) scanf("%d", &vet->array[i]);
}

void showArray(Array* vet) {
        for(int i = 0; i < vet->n; i++) {
                printf("%d ", vet->array[i]);
        }
        printf("\n");
}

void swap(Array* vet, int i, int j) {
        int aux = vet->array[i];
        vet->array[i] = vet->array[j];
        vet->array[j] = aux;
}
//---
//SELECTION
void selection(Array* vet, int* comp, int* mov) {
        for(int i = 0; i < vet->n-1; i++) {
                int menorIndex = i;
                for(int j = i+1; j < vet->n; j++) {
                        (*comp)++;
                        if(vet->array[j] < vet->array[menorIndex]) menorIndex = j;
                }
                swap(vet, i, menorIndex);
                (*mov) +=3;
        }
}

void showAnalysis(int comp, int mov) {
    printf("Comparacoes : %d\nMovimentacoes: %d\n", comp, mov);
}
//---

int main(void) {

        Array* vet = malloc(sizeof(Array));
        printf("Tamanho do vetor: "); scanf("%d", &vet->n);
        readArray(vet);
        showArray(vet);

        int comp = 0, mov = 0;

        selection(vet, &comp, &mov);
        showArray(vet);
        showAnalysis(comp, mov);

        free(vet);

        return 0;
}

File: Disciplinas/AED/AED2/slides/u03-Ordenacao_Interna/Selection/SelectionAnalysis.java
import java.util.Scanner;

class SelectionAnalysis {

    public static Scanner scan = new Scanner(System.in);

    public static int comp = 0;
    public static int mov = 0;

    public static void selection(Array vet) {
        for(int i = 0; i < vet.n-1; i++) {
            int menorIndex = i;
            for(int j = i+1; j < vet.n; j++) {
                comp++;
                if(vet.array[j] < vet.array[menorIndex]) menorIndex = j;
            }
            vet.swap(i, menorIndex);
            mov += 3;
        }
    }

    public static void showAnalysis() {
        System.out.println("Comparacoes: " + comp + "\nMovimentacoes: " + mov);
    }

    public static void main(String [] args) {

        System.out.print("Tamanho do vetor: ");
        int n = scan.nextInt();
        Array vet = new Array(n);
        vet.showArray();

        selection(vet);
        vet.showArray();
        showAnalysis();

        
    }
}

class Array {

    public static Scanner scan = new Scanner(System.in);

    public int[] array;
    public int n;

    public Array() { this(0); }
    public Array(int n) {
        this.n = n;
        this.array = new int[n];
        System.out.print("Insira os " + n + " elementos: ");
        for(int i = 0; i < n; i++) array[i] = scan.nextInt();
    }

    public void showArray() {
        for(int i = 0; i < n; i++) System.out.print(array[i] + " ");
        System.out.println();
    }

    public void swap(int i, int j) {
        int aux = array[i];
        array[i] = array[j];
        array[j] = aux;
    }
}

